### PR TITLE
Meeting Request

### PR DESCRIPTION
**Pull request by lbourret**
Meeting slot finding algorithm for finding free time for attendees to schedule a meeting. Full description of problem:
https://github.com/googleinterns/step/blob/master/walkthroughs/week-5-tdd/project/calendar-walkthrough.md 

PR fulfills required (find time for all required attendees and all optional attendees if possible) and optional (find time for all required and maximum of optional attendees) portions of problem.

### CODE REVIEW

**Comment by meltsufin** [comment]
I think you need to rebase because a commit from the previous PR got in here.
```
git fetch origin
git rebase origin/master
git push -f -u origin week5
```

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +271,142 @@ public void notEnoughRoom() {
...
+    // Day     : |-------------------------------|
+    // Options :       |-----|
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0830AM, false),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartEnd(TIME_0900AM, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0930AM, 15, false),
+            Arrays.asList(PERSON_B)),
```
Comment (meltsufin) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:352]: Compilation error here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,99 @@
...
+    Collections.sort(events, TimeRange.ORDER_BY_START);
+
+    for (int i = 0; i < events.size() - 1; i++) {
+      TimeRange current = events.get(i);
+      TimeRange next = events.get(i+1);
+      boolean nextEnd = (next.end() == TimeRange.END_OF_DAY);
+
+      // Unable to schedule meeting between events
+      if (current.overlaps(next)) {
+        if ((next.end() >= current.end()) || (next.start() - current.start() < duration)) {
```
Comment (meltsufin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:86]: Interesting that you're also blocking off short gaps. That's a neat idea!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,99 @@
...
+   * Merge all attendees' events
+   */
+  private ArrayList<TimeRange> merge(ArrayList<TimeRange> events, MeetingRequest request) {
+    long duration = request.getDuration();
+    Collections.sort(events, TimeRange.ORDER_BY_START);
+
+    for (int i = 0; i < events.size() - 1; i++) {
+      TimeRange current = events.get(i);
+      TimeRange next = events.get(i+1);
+      boolean nextEnd = (next.end() == TimeRange.END_OF_DAY);
```
Comment (meltsufin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:82]: The name `nextEnd` seems a bit ambiguous.
It seems to handle the end of day edge case, but maybe you can explain what it is in the comments?

**Review by meltsufin — CHANGES_REQUESTED** [review]
The algorithm seems good, although I have not been able to run the tests because there are compilation errors.
Please fix the compilation errors and rebase, as I suggested in the comment above.

Also, you handling of meeting length being great than a day, is totally fine.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,129 @@
...
+    }
+    return options;
+  }
+
+  /**
+   * Block off all unavailable time
+   */
+  private ArrayList<TimeRange> unavailableTime(Collection<Event> allEvents, Collection<String> attendees, long duration) {
+    ArrayList<TimeRange> blockedTimeRanges = findAttendeesBlockedTimeRanges(allEvents, attendees);
+    return mergeTimeRanges(blockedTimeRanges, duration);
```
Comment (elefeint) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:100]: Completely non-actionable, but fun for future algorithmic experiments: reducing input size (by merging blocks in this case) and pre-sorting in general can be a nice optimization, but in this case it will actually slow the code down. Merging, due to the sort required, takes O(n*log(n)), where n = number of busy meetings. Without merging, your algorithm still works perfectly, but it's O(n * m), where `m` is the number of free slots. Observe, however, that `m` trends downward to 0 very quickly as input grows because you remove too-small free slots as soon as you see them.

```
Without calling merge (just straight returning `blockedTimeRanges`):
With 1 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 3
With 10 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 8
With 100 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 10
With 1000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 30
With 10000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 114



With calling merge (code as is):
With 1 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 4
With 10 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 7
With 100 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 12
With 1000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 103
With 10000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 3316
```

**Review by elefeint — DISMISSED** [review]
Your algorithm is great! In fact, at a high level it's the same algorithm that I implemented when testing this problem (except mine is not optimized at all, so mine is about twice slower on large inputs than your version without merge.

**Comment by lbourret** [comment]
I realized I could probably optimize the algorithm if I used the method for optimizing opt attendees (counting the instances of each free time range) for finding free times for required attendees (just make sure all attendees are free for each time range), but I got a little tired of this algorithm...

**Review by meltsufin — APPROVED** [review]
Well organized and clean code!
Let's merge this in.
